src/training/graph_group.h

```diff
@@ -112,12 +112,12 @@ class AsyncGraphGroup : public GraphGroup {
           graph->forward();
         }
 
-        if(shardOpt_.size() == 0){
-          for (auto device : devices_) {
+        if(shardOpt_.size() == 0)
+          for (auto device : devices_) 
             shardOpt_.push_back(Optimizer(options_));
 
-          }
-        }
+        shardSync_.reset( new std::mutex[devices_.size()] );
+         
```

Comment:
based on this http://stackoverflow.com/questions/16465633/how-can-i-use-something-like-stdvectorstdmutex

seems working fine